File: OrchidCore/src/main/java/com/eden/orchid/api/resources/resource/OrchidResource.java
package com.eden.orchid.api.resources.resource;

import com.eden.common.json.JSONElement;
import com.eden.common.util.EdenPair;
import com.eden.common.util.EdenUtils;
import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.compilers.OrchidPrecompiler;
import com.eden.orchid.api.theme.pages.OrchidReference;
import org.apache.commons.io.IOUtils;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.Map;

/**
 * Resources provide the "intrinsic content" of a page.
 *
 * @orchidApi extensible
 * @since v1.0.0
 */
public abstract class OrchidResource {

    // TODO: remove this. Resources shouldn't _have_ a resource, but pages should _create_ a reference from APIs on the resource
    protected final OrchidReference reference;

    protected String rawContent;
    protected String content;
    protected JSONElement embeddedData;

    public OrchidResource(OrchidReference reference) {
        if (reference == null) {
            throw new IllegalArgumentException("A resource must have a valid OrchidReference");
        } else {
            this.reference = reference;
        }
    }

    public boolean shouldPrecompile() {
        JSONElement data = getEmbeddedData();
        if (EdenUtils.elementIsObject(data) && ((JSONObject) data.getElement()).has("precompile")) {
            return ((JSONObject) data.getElement()).getBoolean("precompile");
        }

        return reference.getContext().resolve(OrchidPrecompiler.class).shouldPrecompile(reference.getExtension(), getRawContent());
    }

    public boolean shouldRender() {
        return true;
    }

    public InputStream getContentStream() {
        return IOUtils.toInputStream(getRawContent(), StandardCharsets.UTF_8);
    }

    public String compileContent(Object data) {
        if (!EdenUtils.isEmpty(getContent())) {
            String compiledContent = getContent();

            if (shouldPrecompile()) {
                compiledContent = reference.getContext().compile(getPrecompilerExtension(), compiledContent, data);
            }

            return reference.getContext().compile(
                    getReference().getExtension(),
                    compiledContent,
                    data
            );
        } else {
            return "";
        }
    }

    public String getPrecompilerExtension() {
        JSONElement data = getEmbeddedData();
        if (EdenUtils.elementIsObject(data) && ((JSONObject) data.getElement()).has("precompileAs")) {
            return ((JSONObject) data.getElement()).getString("precompileAs");
        }

        return reference.getContext().getDefaultPrecompilerExtension();
    }

    public boolean canUpdate() {
        return false;
    }

    public boolean canDelete() {
        return false;
    }

    public void update(InputStream newContent) throws IOException {

    }

    public void delete() throws IOException {

    }

    @Override
    public String toString() {
        return "OrchidResource[" + this.getClass().getSimpleName() + "]{" +
                "reference=" + reference.getRelativePath() +
                '}';
    }

// Delombok
//----------------------------------------------------------------------------------------------------------------------

    public OrchidReference getReference() {
        return this.reference;
    }

    public String getContent() {
        loadContent();
        return this.content;
    }

    public String getRawContent() {
        loadContent();
        return this.rawContent;
    }

    public JSONElement getEmbeddedData() {
        loadContent();
        return this.embeddedData;
    }

// Freeable Resource Impl
//----------------------------------------------------------------------------------------------------------------------

    protected void loadContent() {
        if(rawContent != null) {
            EdenPair<String, Map<String, Object>> parsedContent = reference.getContext().getEmbeddedData(reference.getExtension(), rawContent);
            this.content = parsedContent.first;
            this.embeddedData = new JSONElement(new JSONObject(parsedContent.second));
        }
        else {
            this.rawContent = "";
            this.content = "";
            this.embeddedData = null;
        }
    }

    public void free() {
        rawContent = null;
        content = null;
    }

}

File: OrchidCore/src/main/kotlin/com/eden/orchid/api/resources/resource/OrchidResource.kt
package com.eden.orchid.api.resources.resource

import com.eden.common.util.EdenUtils
import com.eden.orchid.api.compilers.OrchidPrecompiler
import com.eden.orchid.api.theme.pages.OrchidReference
import com.eden.orchid.utilities.readToString
import java.io.IOException
import java.io.InputStream

abstract class OrchidResource(
    // TODO: remove this. Resources shouldn't _have_ a reference, but pages should _create_ a reference from APIs on the resource
    val reference: OrchidReference
) {

    protected var _content: String? = null
    protected var _embeddedData: Map<String, Any?> = emptyMap()

    abstract fun getContentStream(): InputStream

    open fun shouldPrecompile(): Boolean {
        val data = embeddedData
        if (data.containsKey("precompile")) {
            return data["precompile"].toString().toBoolean()
        }

        val rawContent = getContentStream().readToString()
        return reference
            .context
            .resolve(OrchidPrecompiler::class.java)
            .shouldPrecompile(reference.extension, rawContent)
    }

    open fun shouldRender(): Boolean {
        return true
    }

    open fun compileContent(data: Any?): String {
        var compiledContent = content
        return if (!EdenUtils.isEmpty(compiledContent)) {
            if (shouldPrecompile()) {
                compiledContent = reference.context.compileWithSourceObject(this, precompilerExtension, compiledContent, data)
            }
            reference.context.compileWithSourceObject(
                this,
                reference.extension,
                compiledContent,
                data
            )
        } else {
            ""
        }
    }

    open val precompilerExtension: String
        get() {
            val data = embeddedData
            return if (data.containsKey("precompileAs")) {
                data["precompileAs"].toString()
            } else reference.context.defaultPrecompilerExtension
        }

    open fun canUpdate(): Boolean {
        return false
    }

    open fun canDelete(): Boolean {
        return false
    }

    @Throws(IOException::class)
    open fun update(newContent: InputStream) {
    }

    @Throws(IOException::class)
    open fun delete() {
    }

    override fun toString(): String {
        return "OrchidResource[" + this.javaClass.simpleName + "]{" + "reference=" + reference.relativePath + '}'
    }

    open val content: String
        get() {
            loadContent()
            return _content ?: ""
        }

    open val embeddedData: Map<String, Any?>
        get() {
            loadContent()
            return _embeddedData
        }

    // Freeable Resource Impl
//----------------------------------------------------------------------------------------------------------------------
    protected fun loadContent() {
        val rawContent = getContentStream().readToString()
        if (rawContent != null) {
            val parsedContent = reference.context.getEmbeddedData(reference.extension, rawContent)
            _content = parsedContent.first
            _embeddedData = parsedContent.second ?: emptyMap()
        } else {
            _content = ""
            _embeddedData = emptyMap()
        }
    }

    open fun free() {
        _content = null
        _embeddedData = emptyMap()
    }
}
